apis/Google.Cloud.Firestore.Data/Google.Cloud.Firestore.Data/WriteBatch.cs

```diff
@@ -170,6 +168,93 @@ public WriteBatch Set(DocumentReference documentReference, object documentData,
         }
 
+        /// <summary>
+        /// Adds a Write with an Update field to the list of writes, having first removed all sentinel values.
+        /// If any sentinel values are server timestamps, they are transformed into a new write, which is also added.
+        /// All methods creating an update should call this method rather than calling Writes.Add(write) themselves.
+        /// </summary>
+        private void AddUpdate(Write writeWithUpdate)
+        {
+            var transform = new DocumentTransform { Document = writeWithUpdate.Update.Name };
+            ProcessSentinelValues(writeWithUpdate.Update.Fields, FieldPath.Empty);
+            Writes.Add(writeWithUpdate);
+            if (transform.FieldTransforms.Count != 0)
+            {
+                Writes.Add(new Write { Transform = transform });
+            }
+
+            // Processes the sentinel values within a map of fields, recursively.
+            // - Dictionary entries will be removed for sentinel fields
+            // - The transform will be updated 
+            void ProcessSentinelValues(IDictionary<string, Value> fields, FieldPath parentPath)
+            {
+                List<string> removals = null;
+                foreach (var pair in fields)
+                {
+                    Value value = pair.Value;
+                    string key = pair.Key;
+                    if (value.IsServerTimestampSentinel())
+                    {
+                        transform.FieldTransforms.Add(new FieldTransform
+                        {
+                            // TODO: FieldPath.Append is always used with a freshly-constructed single segment.
+                            // Let's change it...
+                            FieldPath = parentPath.Append(new FieldPath(key)).EncodedPath,
+                            SetToServerValue = ServerValue.RequestTime
+                        });
+                        removals = AddOrCreate(removals, key);
+                    }
+                    else if (value.IsDeleteSentinel())
+                    {
+                        removals = AddOrCreate(removals, key);
+                    }
+                    else if (value.MapValue != null)
+                    {
+                        ProcessSentinelValues(value.MapValue.Fields, parentPath.Append(new FieldPath(key)));
+                    }
+                    else if (value.ArrayValue != null)
+                    {
+                        ValidateNoSentinelValues(value.ArrayValue.Values);
+                    }
+                }
+                removals?.ForEach(key => fields.Remove(key));
+            }
+
+            // Adds an item to an existing list, or creates a new list containing the item if there's currently no
+            // list. This allows us to avoid creating lots of lists redundantly.
+            List<string> AddOrCreate(List<string> list, string item)
```

Comment:
Make this a nested function within `ProcessSentinelValues`, and you don't need to pass `list` as a parameter ;)